illumos-utils/src/zfs.rs

```diff
@@ -203,14 +207,84 @@ pub struct EncryptionDetails {
 
 #[derive(Debug, Default)]
 pub struct SizeDetails {
-    pub quota: Option<usize>,
-    pub reservation: Option<usize>,
+    pub quota: Option<ByteCount>,
+    pub reservation: Option<ByteCount>,
     pub compression: CompressionAlgorithm,
 }
 
+#[derive(Debug)]
+pub struct DatasetProperties {
+    /// The Uuid of the dataset
+    pub id: Option<DatasetUuid>,
+    /// The full name of the dataset.
+    pub name: String,
+    /// Remaining space in the dataset and descendents.
+    pub avail: ByteCount,
+    /// Space used by dataset and descendents.
+    pub used: ByteCount,
+    /// Maximum space usable by dataset and descendents.
+    pub quota: Option<ByteCount>,
+    /// Minimum space guaranteed to dataset and descendents.
+    pub reservation: Option<ByteCount>,
+    /// The compression algorithm used for this dataset.
+    pub compression: String,
+}
+
+impl FromStr for DatasetProperties {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut iter = s.split_whitespace();
+
+        let id = match iter.next().context("Missing UUID")? {
+            "-" => None,
+            anything_else => Some(anything_else.parse::<DatasetUuid>()?),
+        };
+
+        let name = iter.next().context("Missing 'name'")?.to_string();
```

Comment:
1. Is the format documented somewhere? (saw there were tests for it below). If so, could you add a link to it here, otherwise a comment explaining the possibilities?
2. This is fine as is, but if it gets more complicated, then a parser library (my fav is winnow) would be a really good fit.

Reply:
Added it as a constant, as you recommended